pkg/controller/namespace/namespace_controller.go

```diff
@@ -0,0 +1,165 @@
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileNamespace implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileNamespace{}
+
+// ReconcileNamespace reconciles a Namespace object
+type ReconcileNamespace struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a Namespace object and makes changes based on the state read
+// and what is in the Namespace.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  This example creates
+// a Pod as an example
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileNamespace) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling Namespace")
+
+	// Fetch the Namespace instance
+	instance := &corev1.Namespace{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	namespaceList := &corev1.NamespaceList{}
+	opts := client.ListOptions{Namespace: request.Namespace}
+	err = r.client.List(context.TODO(), &opts, namespaceList)
+	if err != nil {
+		reqLogger.Error(err, "Failed to get namespaceList")
+		return reconcile.Result{}, err
+	}
+
+	subjectPermissionList := &managedv1alpha1.SubjectPermissionList{}
+	opts = client.ListOptions{Namespace: request.Namespace}
+	err = r.client.List(context.TODO(), &opts, subjectPermissionList)
+	if err != nil {
+		reqLogger.Error(err, "Failed to get clusterRoleBindingList")
+		return reconcile.Result{}, err
+	}
+
+	clusterRoleList := &v1.ClusterRoleList{}
+	opts = client.ListOptions{Namespace: request.Namespace}
+	err = r.client.List(context.TODO(), &opts, clusterRoleList)
+	if err != nil {
+		reqLogger.Error(err, "Failed to get clusterRoleList")
+		return reconcile.Result{}, err
+	}
+
+	// loop through all subject permissions
+	for _, subjectPermission := range subjectPermissionList.Items {
+		// loop through all permissions in each
+		for _, permission := range subjectPermission.Spec.Permissions {
+			// 1st pass, appy allow regex
+			sl := controllerutil.AllowedNamespacesList(permission.NamespacesAllowedRegex, namespaceList)
```

Comment:
Similar to the remark about the outer loop comments: Why is it important to have this?

I can see, down on lines 123 and 124 that this is the first half of the allowFirst design pattern, but maybe it would help if these two halves were put into their own function that could be called here? Something like: `safeList := controllerutil.GenerateSafeList(permission.NamespacesAllowRegex,permission.NamespaceDeniedRegex,namespaceList`? In that function you could even call the first pass and second pass functions independently to reuse what you have here.

Reply:
I agree. Will move this to util.

Reply:
+1 to this.